utils/stream/stream.go

```diff
@@ -0,0 +1,189 @@
+	return FromReader(bytes.NewReader(b))
+}
+
+// FromBytes streams the given bytes as a buffer.
+func FromString(s string) Source {
+	return FromBytes([]byte(s))
+}
+
+// FromError returns a nil ReadCloser and the error passed when called.
+func FromError(err error) Source {
+	return func() (io.ReadCloser, error) {
+		return nil, err
+	}
+}
+
+// FromFile joins the segments of the path and reads from it.
+func FromFile(path ...string) Source {
+	return FromReadCloserError(os.Open(filepath.Join(path...)))
+}
+
+// FromReadCloserError reads the contents of the readcloser and takes ownership of closing it.
+// If err is non-nil, it is returned as a source.
+func FromReadCloserError(rc io.ReadCloser, err error) Source {
+	return func() (io.ReadCloser, error) {
+		return rc, err
+	}
+}
+
+// FromReader converts a Reader to a ReadCloser and takes ownership of closing it.
+func FromReader(rc io.Reader) Source {
+	return func() (io.ReadCloser, error) {
+		return ioutil.NopCloser(rc), nil
+	}
+}
+
+// ToFile concatenates the given path segments with filepath.Join, creates any
+// parent directoreis if needed, and writes the file.
+func ToFile(path ...string) Dest {
+	return ToModeFile(0600, path...)
+}
+
+// ToModeFile is like ToFile, but sets the permissions on the created file.
+func ToModeFile(mode os.FileMode, path ...string) Dest {
+	return func() (io.WriteCloser, error) {
+		outputPath := filepath.Join(path...)
+		if err := os.MkdirAll(filepath.Dir(outputPath), 0700|os.ModeDir); err != nil {
+			return nil, err
+		}
+
+		return os.OpenFile(outputPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, mode)
+	}
+}
+
+// ToBuffer buffers the contents of the stream and on Close() calls the callback
+// returning its error.
+func ToBuffer(closeCallback func(*bytes.Buffer) error) Dest {
+	return func() (io.WriteCloser, error) {
+		return &bufferCallback{closeCallback: closeCallback}, nil
+	}
+}
+
+// ToYaml unmarshals the contents of the stream as YAML to the given struct.
+func ToYaml(v interface{}) Dest {
+	return ToBuffer(func(buf *bytes.Buffer) error {
+		return yaml.Unmarshal(buf.Bytes(), v)
+	})
+}
+
+// ToError returns an error when Dest is initialized.
+func ToError(err error) Dest {
+	return func() (io.WriteCloser, error) {
+		return nil, err
+	}
+}
+
+// ToDiscard discards any data written to it.
+func ToDiscard() Dest {
+	return ToWriter(ioutil.Discard)
+}
+
+// ToWriter forwards data to the given writer, this function WILL NOT close the
+// underlying stream so it is safe to use with things like stdout.
+func ToWriter(writer io.Writer) Dest {
+	return ToWriteCloser(NopWriteCloser(writer))
+}
+
+// ToWriteCloser forwards data to the given WriteCloser which will be closed
+// after the copy finishes.
+func ToWriteCloser(w io.WriteCloser) Dest {
+	return func() (io.WriteCloser, error) {
+		return w, nil
+	}
+}
+
+// bufferCallback buffers the results and on close calls the callback.
+type bufferCallback struct {
+	bytes.Buffer
```

Comment:
Sorry for being dense, but where is the buffer being allocated for this. I'm used to seeing an accompanying NewBuffer() call.

Reply:
Specifically I'm wondering about the case in stream_test.go, where bufferCallback{} is instantiated directly and then subsequently written to.

Reply:
This one got me too, it turns out that the zero value of a buffer is the preferred way to initialize it in most cases: https://golang.org/src/bytes/buffer.go?s=14358:14392#L440

Reply:
Gotcha, thanks.